keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1365 @@
+
+This list may be extended in the future as use cases are presented. Keeping it minimal has significant benefits for the performance of any commands that list the parent objects themselves. Specific tools may further limit the types they support, but any tooling designed to identify all ApplySets must consider the exact list set out above.
+
+While a purpose-made cluster-scoped CRD is a logical choice, the specification has no opinion on this, so as to accommodate the many and various choices existing tools have already made. In choosing types to support as parents, tools should consider what permissions their target users typically have on their clusters; for instance, they may not have permissions to install CRDs.
+
+#### Labels and annotations
+
+ApplySets MUST also have a “parent object” in the cluster. The ApplySet object MUST be labeled with:
+
+```yaml
+applyset.k8s.io/id: <applysetkey> # REQUIRED
+```
+
+The `applyset.k8s.io/id` label is what makes the object an ApplySet parent object. Its value MUST match the value of `applyset.k8s.io/part-of` on the member objects (see [ApplySet Member Objects](#labels)), and MUST comply with the naming constraints outlined in [ApplySet Naming](#applyset-naming).
+
+Additionally, ApplySet parents MUST be labelled with:
+
+```yaml
+applyset.k8s.io/tooling: <toolId> # REQUIRED
+```
+
+The value should be something like `kubectl/v1.27` or `helm/v3` or `kpt/v1.0.0`,
+i.e. `<toolname>/<semver>`, and tooling should refuse to mutate ApplySets belonging to other tools.
+For more background and guidance on this topic, see the [interoperability](#tooling-interoperability) section.
+
+ApplySets MAY have an annotation extending their scope, which MUST be respected by all tools if present:
+
+```yaml
+applyset.k8s.io/additional-namespaces: <ns>[,] # OPTIONAL
+```
+
+The `applyset.k8s.io/additional-namespaces` annotation extends the scope of the ApplySet. By default, the scope of an ApplySet is its parent object's namespace. When the parent is cluster-scoped but refers to namespace kinds (see below), or when the set spans multiple namespaces (which is not recommended, but allowed), this annotation can be used to extend the ApplySet's scoped to the listed namespaces. As cross-namespace ApplySets are not particularly encouraged, we do not currently optimize this further.  In particular, we do not specify the GKs per namespace.  We can add more annotations in future should the need arise.
+
+The value of this annotation is a comma-separated list of the names of the namespaces (other than the ApplySet namespace) in which objects are found, for example `default,kube-system,ns1,ns2`. We reserve the empty value. As with `applyset.k8s.io/contains-group-kinds`, this list of namespaces must be sorted alphabetically, and should be minimal (other than transiently during ApplySet mutations).
```

Comment:
Separately, you could make `*` mean "all namespaces" and use it when there are sufficiently many.

Listing once across all namespaces will likely be better performance-wise than 1 list per namespace anyway.

Reply:
Good idea - we can investigate that during alpha.  Two things to note:

(1) the biggest issue isn't performance here (I believe), but rather permissions - we don't want to fail operations because a user doesn't have permission to something irrelevant to the particular apply/prune.  That's particularly acute with CRDs, and less so with namespaces because of the RBAC model.  But yes, we should consider '*' particularly for applysets that are really cluster-scoped

(2) While cross-cluster applysets are possible, feedback from users is that they aren't really a great idea in practice.  Rather users either have an applyset scoped to a single namespace or that comprises of cluster-scoped objects and some objects in a particular namespace (e.g. `foo-system).  So we deliberately are not optimizing for this use case.

Reply:
Another idea to investigate during implementation: we don't have to list/watch exactly the specified objects.  If we see 10 namespaces, we might switch to an all-namespace list (and fallback to per-namespace list in case of permissions issues).  If we 1000 objects in an inventory annotation, we might still do a by-kind list if we see that 200 of them are RoleBindings.  If a future apiserver supports getting multiple objects by name in one query (e.g. a fieldSelector with IN), we might use that instead.  In other words, this doesn't have to dictate the execution strategy.